### ISSUE TITLE
Type guards/narrowing for classes don't work in the negative case

### ISSUE THREAD

**Issue by gracepetryk** [microsoft/pyright#5648]
**Describe the bug**
When type guarding a class object with `if x is C`, the variable is not narrowed in an `else` block or after the if statement if `x` is reassigned.

**Code or Screenshots**
Inference doesn't work with classes:
```python
from __future__ import annotations

class C:
    pass

def return_class(return_class: bool) -> int | type[C]:
    if return_class:
        return C

    return 3

x = return_class(True)

if x is C:
    reveal_type(x)  # x: type[C] <-- ok
    x = 10
else:
    reveal_type(x)  # x: int | type[C] <-- should be x: int

reveal_type(x)  # x: int | type[C] <-- should be x: int
```

works as expected with instances:
```python
from __future__ import annotations

class C:
    pass

def return_instance(return_instance: bool) -> int | C:
    if return_instance:
        return C()

    return 3

x = return_instance(True)

if isinstance(x, C):
    reveal_type(x)  # x: C <-- ok
    x = 10
else:
    reveal_type(x)  # x: int <-- ok

reveal_type(x)  # x: int <-- ok
```
**VS Code extension or command-line**
LSP in neovim

pyright version: `1.1.320`
python version: `3.9.16` (also occurs in `3.10.12`)

**Comment by erictraut** [comment]
Pyright's behavior here is correct. The annotation `type[C]` represents "the class `C` or a subtype thereof". A type checker cannot assume that if `x is C` evaluates to `False` at runtime that `type[C]` can be eliminated from the original union. 

If you mark class `C` as `@final`, then it is safe for a type checker to assume that no other classes can be derived from `C`, and `type[C]` can then be eliminated.

This isn't an issue for `isinstance` because that call returns `True` for subclasses as well.
